### PR TITLE
Add AddMissingEqualsToTypeDefinition code fixer

### PR DESCRIPTION
**Pull request by cartermp**
Another common error that gets people: https://github.com/dotnet/fsharp/issues/1121

image

Also works for DU and type abbreviation declarations.

image

meow

### CODE REVIEW

**Comment by cartermp** [comment]
This is ready for review

File: vsintegration/src/FSharp.Editor/CodeFix/AddMissingEqualsToTypeDefinition.fs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+    override _.RegisterCodeFixesAsync context : Task =
+        asyncMaybe {
+            let diagnostics =
+                context.Diagnostics
+                |> Seq.filter (fun x -> fixableDiagnosticIds |> Set.contains x.Id)
+                |> Seq.toImmutableArray
+
+            for diagnostic in diagnostics do
+                let message = diagnostic.GetMessage()
```
Comment (KevinRansom) [inline vsintegration/src/FSharp.Editor/CodeFix/AddMissingEqualsToTypeDefinition.fs:30]: Is this really the best we can do?

**Review by KevinRansom — CHANGES_REQUESTED** [review]
looking for '=' in the error text seems a bit too hit or miss.

File: vsintegration/src/FSharp.Editor/CodeFix/AddMissingEqualsToTypeDefinition.fs (hunk 1 above)
Comment (cartermp) [reply vsintegration/src/FSharp.Editor/CodeFix/AddMissingEqualsToTypeDefinition.fs:30]: The error number is already used. This just ensures only these errors that also propose the equals as a fix are used

**Comment by cartermp** [comment]
Okay, this one is fine as-is but a better approach is to make the error code unique rather than filter based on a heuristic

File: vsintegration/src/FSharp.Editor/CodeFix/AddMissingEqualsToTypeDefinition.fs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+                // Ensure it is the "missing '=' in type definition" error.
+                // This is a funky heuristic but it's probably correct.
+                do! Option.guard (message.Contains("'='"))
```
Comment (baronfel) [inline vsintegration/src/FSharp.Editor/CodeFix/AddMissingEqualsToTypeDefinition.fs:33]: I found a counter-example when I was implementing this for ionide:

```fsharp
type Person = { Name = string }
```

in this malformed type declaration, the `=` sign will be an error of code 10, and the message will be something like "Unexpected token '=' in field declaration. yada yada yada".  This message will trigger your current logic and give a poor result. You may want to mate on these two messages specifically for the missing = on the `type Person {` style error and the `type Rabbit ()\n  member _.Blah = ()` style error.

File: vsintegration/src/FSharp.Editor/CodeFix/AddMissingEqualsToTypeDefinition.fs (hunk 1 above)
Comment (cartermp) [reply vsintegration/src/FSharp.Editor/CodeFix/AddMissingEqualsToTypeDefinition.fs:30]: I will exclude FS0010. There can be specialized error messages for this and other cases and I can add separate quick fixes for those.

**Comment by cartermp** [comment]
Okay @KevinRansom I changed this to only trigger on the specific parse error that we raise for type declarations.

I'll likely look into a separate error message for class declarations in a different PR, and then add that to this fixer only when that specific error message is available.

This will avoid the heuristic of checking the contents of the message and avoid the case that @baronfel found.

**Review by KevinRansom — APPROVED** [review]

**Comment by smoothdeveloper** [comment]
@cartermp should those code fixes also come with few tests to at least highlight the happy case?

This will avoid regressions of the happy case, and if later on we find false positive or bugs, having tests for positive case provides aproximate exemple on how to write such test.

I feel this would apply to all the code fixes PR that came recently.

**Pull request merged**